src/analysis/vector/geometry_checker/qgsfeaturepool.cpp

```diff
@@ -58,6 +58,7 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
   {
     // Feature not in cache, retrieve from layer
     // TODO: avoid always querying all attributes (attribute values are needed when merging by attribute)
+    mFeatureSource = std::make_unique<QgsVectorLayerFeatureSource>( mLayer );
```

Comment:
Unfortunately this fix isn't thread safe -- you can't create a feature source from a different thread vs what the layer lives in